### PR TITLE
Workload manifest updates

### PR DESCRIPTION
**Pull request by sfoslund**

### CODE REVIEW

File: src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs
Comment (sfoslund) [inline src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:157]: @dsplaisted @wli3 @joeloff in case you have feedback. 

I haven't been able to test this end to end because some of the specs are conflicting about how to translate between manifest ids/ versions and package ids/versions. For now I'm assuming the package id and version will match the manifest id and version, but let me know if that is not right. Also, none of the existing manifest packages are following any of these patterns so we'll need to sync with them.

Comment (wli3) [reply src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:157]: The latest plan of the record is package version is the same as the manifest version and it is a string ("6.0.1xx")

@mhutch @dsplaisted please update the specs to reflect this decision

**Comment by wli3** [comment]
not done. But I should really finish my PR

File: src/Cli/dotnet/commands/dotnet-workload/install/ManifestVersion.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-workload/install/ManifestVersion.cs:10]: FWIW the manifest reader/ workload resolver logic is using a long as the manifest version, so we'll have to change that throughout.

Comment (wli3) [reply src/Cli/dotnet/commands/dotnet-workload/install/ManifestVersion.cs:10]: https://github.com/dotnet/sdk/pull/16746

is going on that the same time. Could you ask Mikayla to unblock you by checking it in?

**Comment by sfoslund** [comment]
Addressed feedback I've received so far, waiting on workload teams to start generating manifest packages to test with.

**Comment by wli3** [comment]
do you mind to do a rebase in the meantime? It is easier to see the change

File: src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs
Comment (dsplaisted) [inline src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs:85]: How does this behave if there are already files there?  It should delete the existing files and replace them, I believe.

File: src/Cli/dotnet/commands/dotnet-workload/install/IWorkloadManifestUpdater.cs
Comment (dsplaisted) [inline src/Cli/dotnet/commands/dotnet-workload/install/IWorkloadManifestUpdater.cs:9]: This is part of the IAL, does it need to be a separate interface?  Shouldn't these just be methods on `IInstaller`?

File: src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs
Comment (dsplaisted) [inline src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:69]: It looks like this package downloader is used just for the manifest updater but not for the workload installer (even though the workload installer also uses a package downloader).  That seems confusing.  Probably this would be resolved if the manifest updater were merged with the installer (though I think the paths they use are different, so maybe they'd still be separate instances).

File: src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs
Comment (dsplaisted) [inline src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs:37]: Why do the methods on this class take a `featureBand` parameter?  I think the workload manifest provider only provides access to the manifests for the current feature band.  So if you were to pass in a different `featureBand` to any of these methods I don't think things will work.

Is there any scenario where we'd be trying to do operations on a different `featureBand`?

File: src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs
Comment (dsplaisted) [inline src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:157]: Instead of downloading the NuGet package (again) here, can we just copy the files from the advertised manifest instead?  Are there scenarios where we would be installing a manifest that wasn't advertised?

**Review by dsplaisted — COMMENTED** [review]
So far I've just looked at the implementation, but not the test updates.

File: src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs
Comment (wli3) [inline src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:186]: nit: this will cause a concurrent issue if there are 2 dotnet process running at the sametime. I prefer try out best to delete the empty folder. If not, the system should be able to tolerate empty folder.

File: src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs
Comment (wli3) [inline src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:203]: nit: "Read after write race condition". Although we have this all over SDK. The thread safe way is just to delete and catch the exception. Although that would also be "exception as control flow" for some standard. ¯\\_(ツ)_/¯. In conclusion this is ok

File: src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs
Comment (wli3) [reply src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs:37]: In VSMac we need that

File: src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs
Comment (wli3) [inline src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:221]: check the doc of TransactionalAction. _workloadInstaller.InstallWorkloadManifest has a layer of rollback already. You should not need to say it here

Comment (wli3) [reply src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:221]: or write a small program to try out

File: src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs
Comment (wli3) [reply src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs:84]: let's have an explicit flag for preview. I prefer less behavior change between preview vs non preview sdk

File: src/Tests/dotnet-workload-install.Tests/GivenNetSdkManagedWorkloadInstall.cs
Comment (wli3) [inline src/Tests/dotnet-workload-install.Tests/GivenNetSdkManagedWorkloadInstall.cs:65]: nit: there should be less or prefer not branching (if-else) in test, or you could have bug it test code itself. In this case better to just split into 2 tests

File: src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs:90]: Yes, it's intentional at least for now. Not all of the workload teams are producing manifest packages that we can use to update so several of the manifest update operations will always fail.

File: src/Cli/dotnet/commands/dotnet-workload/install/IWorkloadManifestUpdater.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-workload/install/IWorkloadManifestUpdater.cs:9]: This interface manages the advertising manifest, so it's not part of the IAL as it's the same across all installation types.

File: src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs:85]: Yes this should overwrite existing files.

File: src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:69]: Right now they have separate nuget package downloaders because they install to different locations. We could definitely change the download api so that we can reuse the same downloader though.

File: src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:186]: What do you mean by try our best to delete the empty folder? Move this to the try clause?

Comment (wli3) [reply src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:186]: the delete could fail, catch the exception and swallow it

Comment (wli3) [reply src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:186]: put delete in to try part instead of finally

**Comment by sfoslund** [comment]
@wli3 @dsplaisted I think I'm caught up to feedback now. I'm going to start testing end to end with the new manifest packages but in the meantime feel free to add more feedback.

**Comment by sfoslund** [comment]
I've finished testing locally and I believe everything is good to go. A couple of notes:
- Since workload teams have not started producing properly formatted packages I tested by creating my own and using a local package source. Ideally I'd like to test end to end with real packages before this is merged. 
- Manifest updates will only work for workloads that have properly formatted packages (id: {manifest id}.manifest-6.0.100, version matching manifest version, and workload manifest/targets in a data directory inside of the package)
- If installation fails and we have to roll back, we need to be able to re-install the baseline manifest. As a result, we'll need workload teams to have the baseline manifest properly formatted and available as well as any new updated manifests.

**Review by wli3 — APPROVED** [review]

**Comment by sfoslund** [comment]
@jonathanpeppers how soon will properly named and versioned manifest packages be ready? Is it worth waiting on them to merge this PR or should we proceed with the testing I've done locally?

@dsplaisted let me know if you want me to port this to preview 4, for now I'm going to try to merge to main before there are any more conflicts.

**Comment by jonathanpeppers** [comment]
@sfoslund I started on the Android side: https://github.com/xamarin/xamarin-android/pull/5898

It will probably be next week some time before we have the 4 Apple workloads renamed, and these changes will be in Preview 5. Assuming this PR wouldn't break anything, you don't have to wait on us for this.

File: src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs
Comment (dsplaisted) [inline src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs:102]: It sounds like the failure we would expect is if there are no manifest packages available on the NuGet feed.  I think we should check for that explicitly and treat it not as an error (ie when we call DownloadPackageAsync).  I think other failures are probably not expected and we may want to fail.

File: src/Tests/dotnet-workload-install.Tests/GivenDotnetWorkloadInstall.cs
Comment (dsplaisted) [inline src/Tests/dotnet-workload-install.Tests/GivenDotnetWorkloadInstall.cs:101]: Nit: Indentation

**Review by dsplaisted — COMMENTED** [review]
@sfoslund This is already merged but I went back and took another look at it since I hadn't had a chance to do so before.  There's a few (non-critical) things I think would be nice to follow up on and fix.

File: src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:237]: I'm not sure why github is displaying it like that, it seems fine in main: https://github.com/dotnet/sdk/blob/main/src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs#L237

File: src/Tests/dotnet-workload-install.Tests/GivenDotnetWorkloadInstall.cs
Comment (sfoslund) [reply src/Tests/dotnet-workload-install.Tests/GivenDotnetWorkloadInstall.cs:101]: Same here: https://github.com/dotnet/sdk/blob/cb5b3ff37ec316dd455ad6ec78b5a62f7ec6f28a/src/Tests/dotnet-workload-install.Tests/GivenDotnetWorkloadInstall.cs#L101

**Pull request merged**
